api/query/search.go

```diff
@@ -65,140 +67,130 @@ func (r byName) Swap(i, j int)      { r[i], r[j] = r[j], r[i] }
 func (r byName) Less(i, j int) bool { return r[i].Name < r[j].Name }
 
 type readable interface {
-	Get(string) ([]byte, error)
+	NewReader(string) (io.Reader, error)
 }
 
 type readWritable interface {
 	readable
-	Put(string, []byte) error
+	NewWriteCloser(string) (io.WriteCloser, error)
 }
 
 type httpReadable struct {
 	ctx context.Context
 }
 
-func (hr httpReadable) Get(url string) ([]byte, error) {
-	client := urlfetch.Client(hr.ctx)
-	r, err := client.Get(url)
+func (hr httpReadable) NewReader(url string) (io.Reader, error) {
+	r, err := http.Get(url)
 	if err != nil {
 		return nil, err
 	}
-	defer r.Body.Close()
 
-	if r.StatusCode != 200 {
+	if r.StatusCode != http.StatusOK {
 		return nil, fmt.Errorf("Unexpected status code from %s: %d", url, r.StatusCode)
 	}
 
-	var data []byte
-	if data, err = ioutil.ReadAll(r.Body); err != nil {
-		return nil, err
-	}
-
-	return data, nil
+	return r.Body, nil
 }
 
 type gzipReadWritable struct {
 	delegate readWritable
 }
 
-func (gz gzipReadWritable) Get(key string) ([]byte, error) {
-	zipped, err := gz.delegate.Get(key)
+func (gz gzipReadWritable) NewReader(id string) (io.Reader, error) {
+	r, err := gz.delegate.NewReader(id)
 	if err != nil {
 		return nil, err
 	}
-
-	return gunzipData(zipped)
+	return gzip.NewReader(r)
 }
 
-func gunzipData(zipped []byte) ([]byte, error) {
-	gzReader, err := gzip.NewReader(bytes.NewReader(zipped))
+func (gz gzipReadWritable) NewWriteCloser(id string) (io.WriteCloser, error) {
+	w, err := gz.delegate.NewWriteCloser(id)
 	if err != nil {
 		return nil, err
 	}
-	defer gzReader.Close()
-
-	return ioutil.ReadAll(gzReader)
+	return gzip.NewWriter(w), nil
 }
 
-func (gz gzipReadWritable) Put(key string, unzipped []byte) error {
-	var buf bytes.Buffer
-	gzipData(unzipped, &buf)
-	return gz.delegate.Put(key, buf.Bytes())
+type memcacheReadWritable struct {
+	ctx context.Context
 }
 
-func gzipData(unzipped []byte, zipped *bytes.Buffer) error {
-	bufWriter := bufio.NewWriter(zipped)
-	defer bufWriter.Flush()
-	gzWriter := gzip.NewWriter(bufWriter)
-	defer gzWriter.Close()
-
-	_, err := gzWriter.Write(unzipped)
-	return err
+type memcacheWriteCloser struct {
+	memcacheReadWritable
+	key      string
+	b        bytes.Buffer
+	isClosed bool
 }
 
-type memcacheReadWritable struct {
-	ctx context.Context
-}
+var errMemcacheReaderReadAfterClose = errors.New("memcacheReader: Read() after Close()")
 
-func (mc memcacheReadWritable) Get(key string) ([]byte, error) {
+func (mc memcacheReadWritable) NewReader(key string) (io.Reader, error) {
 	item, err := memcache.Get(mc.ctx, key)
-	if item == nil {
+	if err != nil {
 		return nil, err
 	}
+	return bytes.NewReader(item.Value), nil
```

Comment:
Oh, that's just in case if you also want to make memcache a `ReadCloser`. `bytes.Reader` doesn't have `Close()`; that wrapper will give you a no-op one.